numba/unicode.py

```diff
@@ -498,6 +498,67 @@ def find_impl(a, b):
         return find_impl
 
 
+@overload_method(types.UnicodeType, 'count')
+def unicode_count(src, sub, start=None, end=None):
+
+    if not (start is None or isinstance(start, (types.Omitted, types.Integer, types.Optional))):
+        raise TypingError("Start arg must be of type Integer, Omitted or Optional")
+
+    if isinstance(sub, types.UnicodeType):
+        def count_impl(src, sub, start=start, e=end):
+            count = 0
+            src_len = len(src)
+            sub_len = len(sub)
+
+            if start is None:           #This means no optional arguments are given
+                begin = 0
+                new_end = src_len
+            else:
+                begin = start
+                new_end = e
+            if (begin < 0 and new_end < 0 and begin < new_end):
```

Comment:
The next section deals with normalizing and clipping the `start`  and `end` arguments. For simplicity you should probably do that first for each argument separately and you can define a helper function to do so. So for example, if the start argument is negative, take `src_length` and add the negative argument to it, if it then is still negative , clip it to zero. If it is positive on the other hand and beyond/larger than `src_len`, clip it to `src_len`. A similar approach can be taken for the `end` argument, and using the helper function will help to reduce overall code length.  After this, you will have a valid range/slice in which you will search for the substring. At this stage you can and should also check for early exit conditions, such as the range/slice being zero, the length of the substring being longer than the range/slice and so on (there are a few more). Then, if everything is fine, commence with the substring search proper. This approach should also help reduce the number of cascaded `if` statements and make the code more readable overall.

Reply:
I understand your suggestion and I have tried to make some changes to consider that. But the issue is that the algorithm is complex and less readable because of the slicing, bound checking conditions and python characteristic to give an answer without error even if the bounds(start,end args) are not correct or feasible for a range. Another issue with creating a helper function is that if you look at the algorithm carefully you see there are certain conditions that need to be checked in the main function for slicing or for returning count value and this complicates separating out some code into a helper function. So I feel that instead of changing the algorithm I can add more comments to make the code readable. Let me know what you think about that.

Reply:
I think the implementation could work as follows:
1. Normalise the slice starts and ends to concrete integer values
2. Check if the slice start is in bounds else return 0
3. Slice the input string based on the normalised slice
4. Deal with empty input string and empty substring cases
5. Walk the sliced string using a sliding window of the length of the substring (use a slice?!) to check for matches and either bump the window start position by 1 (no match) or window length (match), accumulate the match count and return this.

Hope this helps.

Reply:
@stuartarchibald I have tried to simplify the code to make it more readable in my new changes. Can you review that and see if that looks good? If it doesn't then I'll try simplifying using the algorithm you suggested.